File: src/execution/runtimeSchedulerEngine.ts
/**
 * Converted from Python: core/execution/runtime_scheduler_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function scheduleRuntimeExecution(actions: any, priorities: any = null, cooldown_ticks: any = 0, worker_id: any = "worker:0", tick: any = 0): any {
  priorities = py.or2(priorities, () => ({}));
  var scheduled: any[] = [];
  var index: any;
  var action: any;
  for ([index, action] of py.enumerate(py.slice(actions, null, 10000))) {
    var action_id: any = py.toStr(py.get(action, "id", `action:${py.toStr(index)}`));
    var priority: any = py.toInt(py.get(priorities, action_id, 0));
    py.listAppend(scheduled, {"action": action, "priority": priority, "worker_id": worker_id, "tick": py.add(tick, py.mul(index, py.max([cooldown_ticks, 0]))), "retry": 0, "paced": true});
  }
  scheduled = py.sorted(scheduled, {key: ((item: any) => [(-py.at(item, "priority")), py.at(item, "tick"), py.toStr(py.get(py.at(item, "action"), "id", ""))]) as (item: any) => any});
  return {"scheduled": scheduled, "worker_id": worker_id, "cooldown_ticks": cooldown_ticks, "deterministic": true, "bounded": true};
}

File: lib/src/execution/runtime_scheduler_engine.dart
/// Port of core/execution/runtime_scheduler_engine.schedule_runtime_execution.
Map<String, dynamic> scheduleRuntimeExecution(
  List<Map<String, dynamic>> actions, {
  Map<String, int>? priorities,
  int cooldownTicks = 0,
  String workerId = 'worker:0',
  int tick = 0,
}) {
  final Map<String, int> prio = priorities ?? <String, int>{};
  final List<Map<String, dynamic>> scheduled = <Map<String, dynamic>>[];

  final List<Map<String, dynamic>> bounded =
      actions.length > 10000 ? actions.sublist(0, 10000) : actions;
  final int cooldown = cooldownTicks > 0 ? cooldownTicks : 0;

  for (int index = 0; index < bounded.length; index++) {
    final Map<String, dynamic> action = bounded[index];
    final String actionId = '${action['id'] ?? 'action:$index'}';
    final int priority = prio[actionId] ?? 0;
    scheduled.add(<String, dynamic>{
      'action': action,
      'priority': priority,
      'worker_id': workerId,
      'tick': tick + (index * cooldown),
      'retry': 0,
      'paced': true,
    });
  }

  scheduled.sort((Map<String, dynamic> a, Map<String, dynamic> b) {
    // key = (-priority, tick, str(action.id))
    final int p = (-(a['priority'] as int)).compareTo(-(b['priority'] as int));
    if (p != 0) return p;
    final int t = (a['tick'] as int).compareTo(b['tick'] as int);
    if (t != 0) return t;
    final String ia = '${(a['action'] as Map<String, dynamic>)['id'] ?? ''}';
    final String ib = '${(b['action'] as Map<String, dynamic>)['id'] ?? ''}';
    return ia.compareTo(ib);
  });

  return <String, dynamic>{
    'scheduled': scheduled,
    'worker_id': workerId,
    'cooldown_ticks': cooldownTicks,
    'deterministic': true,
    'bounded': true,
  };
}
